File: FeatureEngineering/run_feature_engineering_moderate.R
options(stringsAsFactors = FALSE)

library(data.table)

bench_dir <- normalizePath(file.path(getwd(), "FeatureEngineering"), mustWork = FALSE)
if (!dir.exists(bench_dir)) {
  args <- commandArgs(trailingOnly = FALSE)
  file_arg <- args[grep("--file=", args)[1L]]
  bench_dir <- normalizePath(dirname(sub("^--file=", "", file_arg)), mustWork = FALSE)
}

timestamp <- format(Sys.time(), "%Y%m%d_%H%M%S")
out_dir <- Sys.getenv("FEATURE_ENGINEERING_BENCHMARK_OUTPUT", unset = file.path(bench_dir, "outputs", paste0("moderate_", timestamp)))
dir.create(out_dir, recursive = TRUE, showWarnings = FALSE)

rodeo_root <- normalizePath(file.path(getwd(), "..", "Rodeo"), mustWork = FALSE)
if (!dir.exists(rodeo_root)) rodeo_root <- "C:/Users/Bizon/Documents/GitHub/Rodeo"
source(file.path(rodeo_root, "R", "FeatureEngineering_vNext.R"))

legacy_loaded <- FALSE
try({
  source(file.path(rodeo_root, "R", "FeatureEngineering_NumericTypes.R"))
  source(file.path(rodeo_root, "R", "FeatureEngineering_CharacterTypes.R"))
  source(file.path(rodeo_root, "R", "FeatureEngineering_CalendarTypes.R"))
  legacy_loaded <- TRUE
}, silent = TRUE)

rows_grid <- as.integer(strsplit(Sys.getenv("FE_BENCH_ROWS", "10000,100000,500000"), ",")[[1L]])
card_grid <- c(low = 5L, medium = 50L, high = 500L)
shape_grid <- list(
  narrow = list(numeric = 5L, categorical = 2L, date = 1L, text = 1L),
  medium = list(numeric = 20L, categorical = 5L, date = 2L, text = 3L),
  wide = list(numeric = 100L, categorical = 10L, date = 2L, text = 5L)
)
families <- c("numeric", "categorical", "calendar", "text", "interactions", "missingness", "combined_plan")
max_generated_cells <- as.numeric(Sys.getenv("FE_BENCH_MAX_GENERATED_CELLS", "80000000"))

make_data <- function(n, shape, cardinality) {
  set.seed(1000L + n + cardinality + shape$numeric)
  dt <- data.table(id = seq_len(n))
  for (j in seq_len(shape$numeric)) {
    set(dt, j = paste0("num", j), value = runif(n, 0, 100) + j)
  }
  levels <- paste0("L", seq_len(cardinality))
  for (j in seq_len(shape$categorical)) {
    set(dt, j = paste0("cat", j), value = sample(levels, n, TRUE))
  }
  for (j in seq_len(shape$date)) {
    set(dt, j = paste0("date", j), value = as.Date("2024-01-01") + sample(0:730, n, TRUE))
  }
  text_values <- c("Hello WORLD", "two words", "ABC123!", "small text sample", "", NA_character_)
  for (j in seq_len(shape$text)) {
    set(dt, j = paste0("text", j), value = sample(text_values, n, TRUE))
  }
  dt
}

feature_estimate <- function(family, shape, cardinality) {
  top_n <- min(cardinality, 10L)
  switch(
    family,
    numeric = shape$numeric * 4L,
    categorical = shape$categorical * (top_n + 2L),
    calendar = shape$date * 5L,
    text = shape$text * 4L,
    interactions = min(20L, max(1L, shape$numeric - 1L) + top_n + shape$categorical),
    missingness = shape$numeric + shape$categorical + shape$text,
    combined_plan = shape$numeric * 4L + shape$categorical * (top_n + 2L) + shape$date * 5L + shape$text * 4L +
      min(20L, max(1L, shape$numeric - 1L) + top_n + shape$categorical) + shape$numeric + shape$categorical + shape$text,
    0L
  )
}

make_plan <- function(family, data, shape, cardinality) {
  num_cols <- paste0("num", seq_len(shape$numeric))
  cat_cols <- paste0("cat", seq_len(shape$categorical))
  date_cols <- paste0("date", seq_len(shape$date))
  text_cols <- paste0("text", seq_len(shape$text))
  top_n <- min(cardinality, 10L)
  empty <- rodeo_feature_plan()
  if (family == "numeric") {
    return(rodeo_feature_plan(numeric = list(columns = num_cols, transforms = c("log1p", "sqrt", "standardize", "winsorize"), winsorize_probs = c(0.01, 0.99))))
  }
  if (family == "categorical") {
    return(rodeo_feature_plan(categorical = list(columns = cat_cols, top_n = top_n, rare_level = "__RARE__", unseen_level = "__UNSEEN__", one_hot = TRUE, keep_original = TRUE)))
  }
  if (family == "calendar") {
    return(rodeo_feature_plan(calendar = list(columns = date_cols, features = c("year", "month", "wday", "quarter", "is_weekend"))))
  }
  if (family == "text") {
    return(rodeo_feature_plan(text = list(columns = text_cols, features = c("char_count", "word_count", "digit_count", "blank"))))
  }
  if (family == "missingness") {
    return(rodeo_feature_plan(missingness = list(columns = c(num_cols, cat_cols, text_cols), suffix = "_is_missing")))
  }
  if (family == "interactions") {
    pairs <- Map(c, num_cols[-length(num_cols)], num_cols[-1L])
    return(rodeo_feature_plan(interactions = list(
      numeric_pairs = pairs,
      categorical_numeric = list(list(categorical = cat_cols[1L], numeric = num_cols[1L])),
      categorical_pairs = if (length(cat_cols) >= 2L) list(cat_cols[1:2]) else list(),
      max_features = 20L
    )))
  }
  if (family == "combined_plan") {
    pairs <- Map(c, num_cols[-length(num_cols)], num_cols[-1L])
    return(rodeo_feature_plan(
      numeric = list(columns = num_cols, transforms = c("log1p", "sqrt", "standardize", "winsorize"), winsorize_probs = c(0.01, 0.99)),
      categorical = list(columns = cat_cols, top_n = top_n, rare_level = "__RARE__", unseen_level = "__UNSEEN__", one_hot = TRUE, keep_original = TRUE),
      calendar = list(columns = date_cols, features = c("year", "month", "wday", "quarter", "is_weekend")),
      text = list(columns = text_cols, features = c("char_count", "word_count", "digit_count", "blank")),
      missingness = list(columns = c(num_cols, cat_cols, text_cols), suffix = "_is_missing"),
      interactions = list(numeric_pairs = pairs, categorical_numeric = list(list(categorical = cat_cols[1L], numeric = num_cols[1L])), categorical_pairs = if (length(cat_cols) >= 2L) list(cat_cols[1:2]) else list(), max_features = 20L)
    ))
  }
  empty
}

record_case <- function(engine, family, dt, shape_name, card_name, cardinality, expression, generated_features) {
  gc()
  before <- Sys.time()
  status <- "success"
  error <- ""
  out_cols <- NA_integer_
  output_mb <- NA_real_
  tryCatch({
    result <- force(expression)
    if (is.list(result) && !is.null(result$engineered_data)) result <- result$engineered_data
    out_cols <- ncol(result)
    output_mb <- as.numeric(object.size(result)) / 1024^2
  }, error = function(e) {
    status <<- "error"
    error <<- conditionMessage(e)
  })
  elapsed <- as.numeric(difftime(Sys.time(), before, units = "secs"))
  data.table(
    engine = engine,
    family = family,
    rows = nrow(dt),
    cardinality = card_name,
    cardinality_levels = cardinality,
    shape = shape_name,
    generated_features = generated_features,
    output_columns = out_cols,
    output_mb = round(output_mb, 3),
    elapsed_seconds = round(elapsed, 6),
    status = status,
    error = error
  )
}

skip_case <- function(engine, family, n, shape_name, card_name, cardinality, generated_features, reason) {
  data.table(
    engine = engine,
    family = family,
    rows = n,
    cardinality = card_name,
    cardinality_levels = cardinality,
    shape = shape_name,
    generated_features = generated_features,
    output_columns = NA_integer_,
    output_mb = NA_real_,
    elapsed_seconds = NA_real_,
    status = "skipped",
    error = reason
  )
}

run_direct_set <- function(dt, family, shape, cardinality) {
  out <- copy(dt)
  num_cols <- paste0("num", seq_len(shape$numeric))
  cat_cols <- paste0("cat", seq_len(shape$categorical))
  date_cols <- paste0("date", seq_len(shape$date))
  text_cols <- paste0("text", seq_len(shape$text))
  if (family == "numeric") {
    for (col in num_cols) {
      x <- out[[col]]
      set(out, j = paste0(col, "_log1p"), value = log1p(x))
      set(out, j = paste0(col, "_sqrt"), value = sqrt(x))
      set(out, j = paste0(col, "_standardize"), value = (x - mean(x)) / sd(x))
      set(out, j = paste0(col, "_winsorize"), value = pmin(pmax(x, quantile(x, .01)), quantile(x, .99)))
    }
  } else if (family == "categorical") {
    for (col in cat_cols) {
      levels <- names(sort(table(out[[col]]), decreasing = TRUE))[seq_len(min(cardinality, 10L))]
      for (lvl in levels) set(out, j = paste0(col, "__", make.names(lvl)), value = as.integer(out[[col]] == lvl))
    }
  } else if (family == "calendar") {
    for (col in date_cols) {
      d <- as.POSIXlt(out[[col]])
      set(out, j = paste0(col, "_year"), value = d$year + 1900L)
      set(out, j = paste0(col, "_month"), value = d$mon + 1L)
      set(out, j = paste0(col, "_wday"), value = d$wday + 1L)
      set(out, j = paste0(col, "_quarter"), value = d$mon %/% 3L + 1L)
      set(out, j = paste0(col, "_is_weekend"), value = as.integer((d$wday + 1L) %in% c(1L, 7L)))
    }
  } else if (family == "text") {
    for (col in text_cols) {
      x <- as.character(out[[col]])
      x[is.na(x)] <- ""
      set(out, j = paste0(col, "_char_count"), value = nchar(x))
      set(out, j = paste0(col, "_word_count"), value = lengths(regmatches(x, gregexpr("\\S+", x))))
      set(out, j = paste0(col, "_digit_count"), value = nchar(gsub("\\D", "", x)))
      set(out, j = paste0(col, "_blank"), value = as.integer(!nzchar(trimws(x))))
    }
  } else if (family == "missingness") {
    for (col in c(num_cols, cat_cols, text_cols)) set(out, j = paste0(col, "_is_missing"), value = as.integer(is.na(out[[col]])))
  } else if (family == "interactions") {
    for (j in seq_len(min(shape$numeric - 1L, 10L))) set(out, j = paste0(num_cols[j], "_x_", num_cols[j + 1L]), value = out[[num_cols[j]]] * out[[num_cols[j + 1L]]])
    levels <- unique(out[[cat_cols[1L]]])[seq_len(min(length(unique(out[[cat_cols[1L]]])), 10L))]
    for (lvl in levels) set(out, j = paste0(cat_cols[1L], "__", make.names(lvl), "_x_", num_cols[1L]), value = as.integer(out[[cat_cols[1L]]] == lvl) * out[[num_cols[1L]]])
  } else {
    out <- rodeo_fit_transform_feature_plan(dt, make_plan("combined_plan", dt, shape, cardinality))$engineered_data
  }
  out
}

run_direct_assign <- function(dt, family, shape, cardinality) {
  out <- copy(dt)
  num_cols <- paste0("num", seq_len(shape$numeric))
  if (family == "numeric") {
    for (col in num_cols) {
      out[, paste0(col, "_log1p") := log1p(get(col))]
      out[, paste0(col, "_sqrt") := sqrt(get(col))]
      out[, paste0(col, "_standardize") := (get(col) - mean(get(col))) / sd(get(col))]
    }
  } else {
    out <- run_direct_set(dt, family, shape, cardinality)
  }
  out
}

run_base_vectorized <- function(dt, family, shape) {
  out <- copy(dt)
  num_cols <- paste0("num", seq_len(shape$numeric))
  if (family == "numeric") {
    for (col in num_cols) out[[paste0(col, "_log1p")]] <- log1p(out[[col]])
  } else if (family == "missingness") {
    for (col in names(out)) out[[paste0(col, "_is_missing")]] <- as.integer(is.na(out[[col]]))
  } else {
    stop("base vectorized benchmark only covers numeric and missingness in this runner")
  }
  out
}

run_legacy <- function(dt, family, shape, cardinality) {
  if (!legacy_loaded) stop("legacy functions could not be sourced")
  num_cols <- paste0("num", seq_len(shape$numeric))
  cat_cols <- paste0("cat", seq_len(shape$categorical))
  date_cols <- paste0("date", seq_len(shape$date))
  if (family == "numeric") return(Standardize(copy(dt), ColNames = num_cols, ScoreTable = FALSE))
  if (family == "categorical") return(DummifyDT(copy(dt), cols = cat_cols, TopN = rep(min(cardinality, 10L), length(cat_cols)), KeepFactorCols = TRUE, OneHot = TRUE, ReturnFactorLevels = FALSE))
  if (family == "calendar") return(CreateCalendarVariables(copy(dt), DateCols = date_cols, AsFactor = FALSE, TimeUnits = c("wday", "month", "quarter", "year")))
  stop("No comparable Rodeo legacy benchmark for this family")
}

results <- list()
for (n in rows_grid) {
  for (shape_name in names(shape_grid)) {
    shape <- shape_grid[[shape_name]]
    for (card_name in names(card_grid)) {
      cardinality <- card_grid[[card_name]]
      dt <- make_data(n, shape, cardinality)
      for (family in families) {
        generated <- feature_estimate(family, shape, cardinality)
        too_large <- (n * generated) > max_generated_cells
        engines <- c("Rodeo vNext", "data.table set", "data.table :=", "base vectorized", "Rodeo legacy", "collapse")
        for (engine in engines) {
          if (too_large) {
            results[[length(results) + 1L]] <- skip_case(engine, family, n, shape_name, card_name, cardinality, generated, "generated-cell guardrail")
            next
          }
          if (engine == "base vectorized" && !family %in% c("numeric", "missingness")) {
            results[[length(results) + 1L]] <- skip_case(engine, family, n, shape_name, card_name, cardinality, generated, "base vectorized runner only covers numeric/missingness")
            next
          }
          if (engine == "Rodeo legacy" && !family %in% c("numeric", "categorical", "calendar")) {
            results[[length(results) + 1L]] <- skip_case(engine, family, n, shape_name, card_name, cardinality, generated, "no comparable legacy function in moderate runner")
            next
          }
          if (engine == "collapse" && !(requireNamespace("collapse", quietly = TRUE) && family == "numeric")) {
            results[[length(results) + 1L]] <- skip_case(engine, family, n, shape_name, card_name, cardinality, generated, "collapse benchmark only covers numeric when package is available")
            next
          }
          expr <- switch(
            engine,
            "Rodeo vNext" = rodeo_fit_transform_feature_plan(dt, make_plan(family, dt, shape, cardinality)),
            "data.table set" = run_direct_set(dt, family, shape, cardinality),
            "data.table :=" = run_direct_assign(dt, family, shape, cardinality),
            "base vectorized" = run_base_vectorized(dt, family, shape),
            "Rodeo legacy" = run_legacy(dt, family, shape, cardinality),
            "collapse" = {
              out <- copy(dt)
              for (col in paste0("num", seq_len(shape$numeric))) {
                set(out, j = paste0(col, "_centered"), value = out[[col]] - collapse::fmean(out[[col]]))
              }
              out
            }
          )
          results[[length(results) + 1L]] <- record_case(engine, family, dt, shape_name, card_name, cardinality, expr, generated)
        }
      }
    }
  }
}

summary <- rbindlist(results, fill = TRUE)
fwrite(summary, file.path(out_dir, "r_feature_engineering_moderate_summary.csv"))

writeLines(capture.output(sessionInfo()), file.path(out_dir, "session_info.txt"))
writeLines(c(
  paste("timestamp:", timestamp),
  paste("rows_grid:", paste(rows_grid, collapse = ",")),
  paste("max_generated_cells:", max_generated_cells),
  paste("machine:", Sys.info()[["nodename"]]),
  paste("system:", paste(Sys.info(), collapse = " | "))
), file.path(out_dir, "system_info.txt"))

cat(normalizePath(out_dir, winslash = "/"), "\n")

File: FeatureEngineering/run_feature_engineering_moderate.py
from __future__ import annotations

import csv
import json
import os
import platform
import sys
import time
from datetime import datetime, timedelta
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
OUT_DIR = Path(os.environ.get("FEATURE_ENGINEERING_BENCHMARK_OUTPUT", ROOT / "FeatureEngineering" / "outputs" / f"moderate_{datetime.now().strftime('%Y%m%d_%H%M%S')}"))
OUT_DIR.mkdir(parents=True, exist_ok=True)
POLARSFE_ROOT = ROOT.parent / "polars_feature_engineering"
if str(POLARSFE_ROOT) not in sys.path:
    sys.path.insert(0, str(POLARSFE_ROOT))

ROWS_GRID = [int(x) for x in os.environ.get("FE_BENCH_ROWS", "10000,100000,500000").split(",")]
CARD_GRID = {"low": 5, "medium": 50, "high": 500}
SHAPES = {
    "narrow": {"numeric": 5, "categorical": 2, "date": 1, "text": 1},
    "medium": {"numeric": 20, "categorical": 5, "date": 2, "text": 3},
    "wide": {"numeric": 100, "categorical": 10, "date": 2, "text": 5},
}
FAMILIES = ["numeric", "categorical", "calendar", "text", "interactions", "missingness", "combined_plan"]
MAX_GENERATED_CELLS = int(os.environ.get("FE_BENCH_MAX_GENERATED_CELLS", "80000000"))


def feature_estimate(family, shape, cardinality):
    top_n = min(cardinality, 10)
    if family == "numeric":
        return shape["numeric"] * 4
    if family == "categorical":
        return shape["categorical"] * (top_n + 2)
    if family == "calendar":
        return shape["date"] * 5
    if family == "text":
        return shape["text"] * 4
    if family == "interactions":
        return min(20, max(1, shape["numeric"] - 1) + top_n + shape["categorical"])
    if family == "missingness":
        return shape["numeric"] + shape["categorical"] + shape["text"]
    if family == "combined_plan":
        return (
            shape["numeric"] * 4
            + shape["categorical"] * (top_n + 2)
            + shape["date"] * 5
            + shape["text"] * 4
            + min(20, max(1, shape["numeric"] - 1) + top_n + shape["categorical"])
            + shape["numeric"]
            + shape["categorical"]
            + shape["text"]
        )
    return 0


def make_data(pl, n, shape, cardinality):
    base_date = datetime(2024, 1, 1)
    data = {"id": list(range(1, n + 1))}
    for j in range(1, shape["numeric"] + 1):
        data[f"num{j}"] = [float((i * (j + 3)) % 1000) / 10.0 + j for i in range(n)]
    levels = [f"L{i}" for i in range(1, cardinality + 1)]
    for j in range(1, shape["categorical"] + 1):
        data[f"cat{j}"] = [levels[(i + j) % cardinality] for i in range(n)]
    for j in range(1, shape["date"] + 1):
        data[f"date{j}"] = [base_date + timedelta(days=(i + j) % 730) for i in range(n)]
    text_values = ["Hello WORLD", "two words", "ABC123!", "small text sample", "", None]
    for j in range(1, shape["text"] + 1):
        data[f"text{j}"] = [text_values[(i + j) % len(text_values)] for i in range(n)]
    return pl.DataFrame(data)


def make_plan(PolarsFE, family, shape, cardinality):
    num_cols = [f"num{i}" for i in range(1, shape["numeric"] + 1)]
    cat_cols = [f"cat{i}" for i in range(1, shape["categorical"] + 1)]
    date_cols = [f"date{i}" for i in range(1, shape["date"] + 1)]
    text_cols = [f"text{i}" for i in range(1, shape["text"] + 1)]
    top_n = min(cardinality, 10)
    kwargs = {}
    if family in ("numeric", "combined_plan"):
        kwargs["numeric"] = {"columns": num_cols, "transforms": ["log1p", "sqrt", "standardize", "winsorize"], "winsorize_probs": [0.01, 0.99]}
    if family in ("categorical", "combined_plan"):
        kwargs["categorical"] = {"columns": cat_cols, "top_n": top_n, "rare_level": "__RARE__", "unseen_level": "__UNSEEN__", "one_hot": True, "keep_original": True}
    if family in ("calendar", "combined_plan"):
        kwargs["calendar"] = {"columns": date_cols, "features": ["year", "month", "wday", "quarter", "is_weekend"]}
    if family in ("text", "combined_plan"):
        kwargs["text"] = {"columns": text_cols, "features": ["char_count", "word_count", "digit_count", "blank"]}
    if family in ("missingness", "combined_plan"):
        kwargs["missingness"] = {"columns": num_cols + cat_cols + text_cols, "suffix": "_is_missing"}
    if family in ("interactions", "combined_plan"):
        kwargs["interactions"] = {
            "numeric_pairs": [[num_cols[i], num_cols[i + 1]] for i in range(max(0, len(num_cols) - 1))],
            "categorical_numeric": [{"categorical": cat_cols[0], "numeric": num_cols[0]}],
            "categorical_pairs": [[cat_cols[0], cat_cols[1]]] if len(cat_cols) > 1 else [],
            "max_features": 20,
        }
    return PolarsFE.polars_feature_plan(**kwargs)


def record(rows, engine, family, df, shape_name, card_name, cardinality, generated_features, fn):
    start = time.perf_counter()
    status = "success"
    error = ""
    output_columns = None
    output_mb = None
    try:
        result = fn()
        if isinstance(result, dict) and "engineered_data" in result:
            result = result["engineered_data"]
        output_columns = len(result.columns)
        output_mb = round(result.estimated_size("mb"), 3)
    except Exception as exc:  # noqa: BLE001
        status = "error"
        error = str(exc)
    rows.append(
        {
            "engine": engine,
            "family": family,
            "rows": df.height,
            "cardinality": card_name,
            "cardinality_levels": cardinality,
            "shape": shape_name,
            "generated_features": generated_features,
            "output_columns": output_columns,
            "output_mb": output_mb,
            "elapsed_seconds": round(time.perf_counter() - start, 6),
            "status": status,
            "error": error,
        }
    )


def skip(rows, engine, family, n, shape_name, card_name, cardinality, generated_features, reason):
    rows.append(
        {
            "engine": engine,
            "family": family,
            "rows": n,
            "cardinality": card_name,
            "cardinality_levels": cardinality,
            "shape": shape_name,
            "generated_features": generated_features,
            "output_columns": "",
            "output_mb": "",
            "elapsed_seconds": "",
            "status": "skipped",
            "error": reason,
        }
    )


def run_direct_polars(pl, df, family, shape, lazy=False):
    num_cols = [f"num{i}" for i in range(1, shape["numeric"] + 1)]
    cat_cols = [f"cat{i}" for i in range(1, shape["categorical"] + 1)]
    date_cols = [f"date{i}" for i in range(1, shape["date"] + 1)]
    text_cols = [f"text{i}" for i in range(1, shape["text"] + 1)]
    base = df.lazy() if lazy else df
    exprs = []
    if family == "numeric":
        for col in num_cols:
            exprs.extend([pl.col(col).log1p().alias(f"{col}_log1p"), pl.col(col).sqrt().alias(f"{col}_sqrt"), ((pl.col(col) - pl.col(col).mean()) / pl.col(col).std()).alias(f"{col}_standardize")])
    elif family == "categorical":
        for col in cat_cols:
            levels = df[col].unique().head(10).to_list()
            exprs.extend([(pl.col(col) == level).cast(pl.Int8).alias(f"{col}__{level}") for level in levels])
    elif family == "calendar":
        for col in date_cols:
            d = pl.col(col).cast(pl.Date)
            exprs.extend([d.dt.year().alias(f"{col}_year"), d.dt.month().alias(f"{col}_month"), d.dt.weekday().alias(f"{col}_wday"), d.dt.quarter().alias(f"{col}_quarter"), d.dt.weekday().is_in([6, 7]).cast(pl.Int8).alias(f"{col}_is_weekend")])
    elif family == "text":
        for col in text_cols:
            x = pl.col(col).cast(pl.Utf8).fill_null("")
            exprs.extend([x.str.len_chars().alias(f"{col}_char_count"), x.str.count_matches(r"\S+").alias(f"{col}_word_count"), x.str.count_matches(r"\d").alias(f"{col}_digit_count"), (x.str.strip_chars().str.len_chars() == 0).cast(pl.Int8).alias(f"{col}_blank")])
    elif family == "missingness":
        exprs.extend([pl.col(col).is_null().cast(pl.Int8).alias(f"{col}_is_missing") for col in num_cols + cat_cols + text_cols])
    elif family == "interactions":
        for i in range(min(len(num_cols) - 1, 10)):
            exprs.append((pl.col(num_cols[i]) * pl.col(num_cols[i + 1])).alias(f"{num_cols[i]}_x_{num_cols[i + 1]}"))
    else:
        return run_direct_polars(pl, df, "numeric", shape, lazy=lazy)
    out = base.with_columns(exprs)
    return out.collect() if lazy else out


def main():
    import polars as pl
    import PolarsFE

    rows = []
    for n in ROWS_GRID:
        for shape_name, shape in SHAPES.items():
            for card_name, cardinality in CARD_GRID.items():
                df = make_data(pl, n, shape, cardinality)
                for family in FAMILIES:
                    generated = feature_estimate(family, shape, cardinality)
                    engines = ["PolarsFE vNext", "Polars eager", "Polars lazy", "pandas", "duckdb"]
                    for engine in engines:
                        if n * generated > MAX_GENERATED_CELLS:
                            skip(rows, engine, family, n, shape_name, card_name, cardinality, generated, "generated-cell guardrail")
                            continue
                        if engine == "pandas":
                            try:
                                import pandas as pd  # noqa: F401
                                pdf = df.to_pandas()
                            except Exception as exc:  # noqa: BLE001
                                skip(rows, engine, family, n, shape_name, card_name, cardinality, generated, str(exc))
                                continue
                            if family != "numeric":
                                skip(rows, engine, family, n, shape_name, card_name, cardinality, generated, "pandas runner only covers numeric")
                                continue
                            record(rows, engine, family, df, shape_name, card_name, cardinality, generated, lambda pdf=pdf: pdf.assign(**{f"{col}_log1p": __import__("numpy").log1p(pdf[col]) for col in [f"num{i}" for i in range(1, shape["numeric"] + 1)]}))
                        elif engine == "duckdb":
                            try:
                                import duckdb  # noqa: F401
                            except Exception as exc:  # noqa: BLE001
                                skip(rows, engine, family, n, shape_name, card_name, cardinality, generated, str(exc))
                                continue
                            skip(rows, engine, family, n, shape_name, card_name, cardinality, generated, "duckdb runner reserved for SQL-friendly follow-up")
                        elif engine == "PolarsFE vNext":
                            plan = make_plan(PolarsFE, family, shape, cardinality)
                            record(rows, engine, family, df, shape_name, card_name, cardinality, generated, lambda plan=plan: PolarsFE.polars_fit_transform_feature_plan(df, plan))
                        elif engine == "Polars eager":
                            record(rows, engine, family, df, shape_name, card_name, cardinality, generated, lambda family=family: run_direct_polars(pl, df, family, shape, lazy=False))
                        elif engine == "Polars lazy":
                            record(rows, engine, family, df, shape_name, card_name, cardinality, generated, lambda family=family: run_direct_polars(pl, df, family, shape, lazy=True))

    out_file = OUT_DIR / "python_feature_engineering_moderate_summary.csv"
    with out_file.open("w", newline="", encoding="utf-8") as fh:
        fieldnames = ["engine", "family", "rows", "cardinality", "cardinality_levels", "shape", "generated_features", "output_columns", "output_mb", "elapsed_seconds", "status", "error"]
        writer = csv.DictWriter(fh, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(rows)

    (OUT_DIR / "system_info_python.json").write_text(
        json.dumps({"python": sys.version, "platform": platform.platform(), "machine": platform.machine(), "rows_grid": ROWS_GRID, "max_generated_cells": MAX_GENERATED_CELLS}, indent=2),
        encoding="utf-8",
    )
    print(OUT_DIR.as_posix())


if __name__ == "__main__":
    main()
